File: redline-js/src/wml/revision.ts
/**
 * Revision markup generation for Word documents
 *
 * Creates w:ins (insertion) and w:del (deletion) elements with proper
 * tracking attributes (author, date, id).
 */

import {
  cloneNode,
  getTagName,
  getChildren,
  type XmlNode,
} from '../core/xml';

/**
 * Settings for revision tracking
 */
export interface RevisionSettings {
  /** Author name for tracked changes */
  author: string;
  /** Date for tracked changes (ISO 8601 format) */
  dateTime: string;
}

/**
 * Default revision settings
 */
export const DEFAULT_REVISION_SETTINGS: RevisionSettings = {
  author: 'redline-js',
  dateTime: new Date().toISOString(),
};

// Counter for unique revision IDs
let revisionIdCounter = 1;

/**
 * Reset the revision ID counter (useful for testing)
 */
export function resetRevisionIdCounter(value = 1): void {
  revisionIdCounter = value;
}

/**
 * Get the next unique revision ID
 */
export function getNextRevisionId(): number {
  return revisionIdCounter++;
}

/**
 * Find the maximum w:id value used in an XML node tree.
 * This is used to avoid ID collisions with existing tracked changes.
 * 
 * Scans for w:id attributes on revision elements (w:ins, w:del, w:rPrChange, w:pPrChange, etc.)
 */
export function findMaxRevisionId(nodes: XmlNode | XmlNode[]): number {
  const nodeArray = Array.isArray(nodes) ? nodes : [nodes];
  let maxId = 0;

  function walk(node: XmlNode): void {
    // Check if this node has a w:id attribute
    const attrs = node[':@'] as Record<string, string> | undefined;
    if (attrs) {
      const idStr = attrs['@_w:id'];
      if (idStr !== undefined) {
        const id = parseInt(idStr, 10);
        if (!isNaN(id) && id > maxId) {
          maxId = id;
        }
      }
    }

    // Recursively check children
    for (const child of getChildren(node)) {
      walk(child);
    }
  }

  for (const node of nodeArray) {
    walk(node);
  }

  return maxId;
}

/**
 * Create a w:ins (insertion) element wrapping content
 *
 * @param content The content to wrap (typically w:r elements)
 * @param settings Revision tracking settings
 * @returns The w:ins element
 */
export function createInsertion(
  content: XmlNode | XmlNode[],
  settings: RevisionSettings = DEFAULT_REVISION_SETTINGS
): XmlNode {
  const children = Array.isArray(content) ? content : [content];

  const insNode: XmlNode = {
    'w:ins': children.map(cloneNode),
    ':@': {
      '@_w:author': settings.author,
      '@_w:id': String(getNextRevisionId()),
      '@_w:date': settings.dateTime,
    },
  };

  return insNode;
}

/**
 * Create a w:del (deletion) element wrapping content
 *
 * @param content The content to wrap (typically w:r elements)
 * @param settings Revision tracking settings
 * @returns The w:del element
 */
export function createDeletion(
  content: XmlNode | XmlNode[],
  settings: RevisionSettings = DEFAULT_REVISION_SETTINGS
): XmlNode {
  const children = Array.isArray(content) ? content : [content];

  // For deletions, we need to convert w:t to w:delText
  const convertedChildren = children.map((child) => convertToDeletedContent(cloneNode(child)));

  const delNode: XmlNode = {
    'w:del': convertedChildren,
    ':@': {
      '@_w:author': settings.author,
      '@_w:id': String(getNextRevisionId()),
      '@_w:date': settings.dateTime,
    },
  };

  return delNode;
}

/**
 * Convert content for deletion (w:t -> w:delText)
 *
 * In Word, deleted text uses w:delText instead of w:t.
 * This function recursively converts text elements.
 */
function convertToDeletedContent(node: XmlNode): XmlNode {
  const tagName = getTagName(node);

  // Convert w:t to w:delText
  if (tagName === 'w:t') {
    const children = node['w:t'];
    const attrs = node[':@'];

    const delTextNode: XmlNode = {
      'w:delText': children,
    };
    if (attrs) {
      delTextNode[':@'] = attrs;
    }
    return delTextNode;
  }

  // Recursively convert children
  if (tagName) {
    const children = getChildren(node);
    if (children.length > 0) {
      node[tagName] = children.map((child) => convertToDeletedContent(child));
    }
  }

  return node;
}

/**
 * Create a w:rPrChange element for run property changes
 */
export function createRunPropertyChange(
  originalProperties: XmlNode,
  settings: RevisionSettings = DEFAULT_REVISION_SETTINGS
): XmlNode {
  return {
    'w:rPrChange': [cloneNode(originalProperties)],
    ':@': {
      '@_w:author': settings.author,
      '@_w:id': String(getNextRevisionId()),
      '@_w:date': settings.dateTime,
    },
  };
}

/**
 * Create a w:pPrChange element for paragraph property changes
 */
export function createParagraphPropertyChange(
  originalProperties: XmlNode,
  settings: RevisionSettings = DEFAULT_REVISION_SETTINGS
): XmlNode {
  return {
    'w:pPrChange': [cloneNode(originalProperties)],
    ':@': {
      '@_w:author': settings.author,
      '@_w:id': String(getNextRevisionId()),
      '@_w:date': settings.dateTime,
    },
  };
}

/**
 * Create a run (w:r) element with text content
 */
export function createRun(text: string, properties?: XmlNode): XmlNode {
  const children: XmlNode[] = [];

  if (properties) {
    children.push(cloneNode(properties));
  }

  children.push({
    'w:t': [{ '#text': text }],
    ':@': text.startsWith(' ') || text.endsWith(' ') ? { '@_xml:space': 'preserve' } : undefined,
  });

  return {
    'w:r': children,
  };
}

/**
 * Create a paragraph (w:p) element
 */
export function createParagraph(runs: XmlNode[], properties?: XmlNode): XmlNode {
  const children: XmlNode[] = [];

  if (properties) {
    children.push(cloneNode(properties));
  }

  children.push(...runs.map(cloneNode));

  return {
    'w:p': children,
  };
}

/**
 * Wrap content in an insertion or deletion based on status
 */
export function wrapWithRevision(
  content: XmlNode | XmlNode[],
  status: 'inserted' | 'deleted',
  settings: RevisionSettings = DEFAULT_REVISION_SETTINGS
): XmlNode {
  if (status === 'inserted') {
    return createInsertion(content, settings);
  } else {
    return createDeletion(content, settings);
  }
}

/**
 * Check if a node is a revision element (w:ins or w:del)
 */
export function isRevisionElement(node: XmlNode): boolean {
  const tagName = getTagName(node);
  return tagName === 'w:ins' || tagName === 'w:del';
}

/**
 * Check if a node is an insertion
 */
export function isInsertion(node: XmlNode): boolean {
  return getTagName(node) === 'w:ins';
}

/**
 * Check if a node is a deletion
 */
export function isDeletion(node: XmlNode): boolean {
  return getTagName(node) === 'w:del';
}

/**
 * Check if a node is a format change (w:rPrChange or w:pPrChange)
 */
export function isFormatChange(node: XmlNode): boolean {
  const tagName = getTagName(node);
  return tagName === 'w:rPrChange' || tagName === 'w:pPrChange';
}

/**
 * Revision element tag names that have w:id attributes
 */
const REVISION_ELEMENT_TAGS = new Set([
  'w:ins',
  'w:del',
  'w:rPrChange',
  'w:pPrChange',
  'w:sectPrChange',
  'w:tblPrChange',
  'w:tblGridChange',
  'w:trPrChange',
  'w:tcPrChange',
  'w:cellIns',
  'w:cellDel',
  'w:cellMerge',
  'w:customXmlInsRangeStart',
  'w:customXmlDelRangeStart',
  'w:customXmlMoveFromRangeStart',
  'w:customXmlMoveToRangeStart',
  'w:moveFrom',
  'w:moveTo',
  'w:moveFromRangeStart',
  'w:moveToRangeStart',
  'w:numberingChange',
]);

/**
 * Fix up revision IDs by renumbering all revision elements sequentially.
 * 
 * This is a post-processing step that ensures all w:id attributes on revision
 * elements (w:ins, w:del, w:rPrChange, etc.) are unique and sequential.
 * 
 * This mirrors the C# WmlComparer.FixUpRevisionIds() approach which:
 * 1. Collects all revisions from main doc, footnotes, endnotes
 * 2. Renumbers them sequentially from 1
 * 
 * @param nodes The XML nodes to process (modified in place)
 */
export function fixUpRevisionIds(nodes: XmlNode | XmlNode[]): void {
  const nodeArray = Array.isArray(nodes) ? nodes : [nodes];
  
  // Collect all revision elements
  const revisionElements: XmlNode[] = [];
  
  function collectRevisions(node: XmlNode): void {
    const tagName = getTagName(node);
    
    // Check if this is a revision element with w:id
    if (tagName && REVISION_ELEMENT_TAGS.has(tagName)) {
      const attrs = node[':@'] as Record<string, string> | undefined;
      if (attrs && '@_w:id' in attrs) {
        revisionElements.push(node);
      }
    }
    
    // Recursively check children
    for (const child of getChildren(node)) {
      collectRevisions(child);
    }
  }
  
  // Collect from all nodes
  for (const node of nodeArray) {
    collectRevisions(node);
  }
  
  // Renumber sequentially from 1
  let nextId = 1;
  for (const element of revisionElements) {
    const attrs = element[':@'] as Record<string, string>;
    attrs['@_w:id'] = String(nextId++);
  }
}

/**
 * Count revisions in a document tree
 */
export function countRevisions(nodes: XmlNode | XmlNode[]): {
  insertions: number;
  deletions: number;
  formatChanges: number;
  total: number;
} {
  const nodeArray = Array.isArray(nodes) ? nodes : [nodes];
  let insertions = 0;
  let deletions = 0;
  let formatChanges = 0;

  function walk(node: XmlNode) {
    if (isInsertion(node)) {
      insertions++;
    } else if (isDeletion(node)) {
      deletions++;
    } else if (isFormatChange(node)) {
      formatChanges++;
    }

    for (const child of getChildren(node)) {
      walk(child);
    }
  }

  for (const node of nodeArray) {
    walk(node);
  }

  return {
    insertions,
    deletions,
    formatChanges,
    total: insertions + deletions + formatChanges,
  };
}

File: redline-rs/crates/redline-core/src/wml/revision.rs
use crate::xml::arena::XmlDocument;
use crate::xml::namespaces::{W, W16DU};
use crate::xml::node::XmlNodeData;
use crate::xml::xname::{XAttribute, XName};
use indextree::NodeId;
use std::sync::atomic::{AtomicI32, Ordering};

static REVISION_ID_COUNTER: AtomicI32 = AtomicI32::new(0);

#[derive(Debug, Clone)]
pub struct RevisionSettings {
    pub author: String,
    pub date_time: String,
}

impl Default for RevisionSettings {
    fn default() -> Self {
        Self {
            author: "redline-rs".to_string(),
            date_time: chrono::Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string(),
        }
    }
}

pub fn reset_revision_id_counter(value: i32) {
    REVISION_ID_COUNTER.store(value, Ordering::SeqCst);
}

pub fn get_next_revision_id() -> i32 {
    REVISION_ID_COUNTER.fetch_add(1, Ordering::SeqCst)
}

pub fn find_max_revision_id(doc: &XmlDocument, start: NodeId) -> i32 {
    let mut max_id = 0;
    
    for node_id in doc.descendants(start) {
        if let Some(data) = doc.get(node_id) {
            if let Some(attrs) = data.attributes() {
                for attr in attrs {
                    if attr.name.local_name == "id" && attr.name.namespace.as_deref() == Some(W::NS) {
                        if let Ok(id) = attr.value.parse::<i32>() {
                            max_id = max_id.max(id);
                        }
                    }
                }
            }
        }
    }
    
    max_id
}

pub fn create_insertion(
    doc: &mut XmlDocument,
    parent: NodeId,
    settings: &RevisionSettings,
) -> NodeId {
    let rev_id = get_next_revision_id();
    let attrs = vec![
        XAttribute::new(W::id(), &rev_id.to_string()),  // w:id MUST come first per ECMA-376
        XAttribute::new(W::author(), &settings.author),
        XAttribute::new(W::date(), &settings.date_time),
        // Add w16du:dateUtc for modern Word timezone handling
        XAttribute::new(W16DU::dateUtc(), &settings.date_time),
    ];
    
    doc.add_child(parent, XmlNodeData::element_with_attrs(W::ins(), attrs))
}

pub fn create_deletion(
    doc: &mut XmlDocument,
    parent: NodeId,
    settings: &RevisionSettings,
) -> NodeId {
    let rev_id = get_next_revision_id();
    let attrs = vec![
        XAttribute::new(W::id(), &rev_id.to_string()),  // w:id MUST come first per ECMA-376
        XAttribute::new(W::author(), &settings.author),
        XAttribute::new(W::date(), &settings.date_time),
        // Add w16du:dateUtc for modern Word timezone handling
        XAttribute::new(W16DU::dateUtc(), &settings.date_time),
    ];
    
    doc.add_child(parent, XmlNodeData::element_with_attrs(W::del(), attrs))
}

pub fn create_run_property_change(
    doc: &mut XmlDocument,
    parent: NodeId,
    settings: &RevisionSettings,
) -> NodeId {
    let rev_id = get_next_revision_id();
    let attrs = vec![
        XAttribute::new(W::id(), &rev_id.to_string()),  // w:id MUST come first per ECMA-376
        XAttribute::new(W::author(), &settings.author),
        XAttribute::new(W::date(), &settings.date_time),
    ];
    
    doc.add_child(parent, XmlNodeData::element_with_attrs(W::rPrChange(), attrs))
}

pub fn create_paragraph_property_change(
    doc: &mut XmlDocument,
    parent: NodeId,
    settings: &RevisionSettings,
) -> NodeId {
    let rev_id = get_next_revision_id();
    let attrs = vec![
        XAttribute::new(W::id(), &rev_id.to_string()),  // w:id MUST come first per ECMA-376
        XAttribute::new(W::author(), &settings.author),
        XAttribute::new(W::date(), &settings.date_time),
    ];
    
    doc.add_child(parent, XmlNodeData::element_with_attrs(W::pPrChange(), attrs))
}

pub fn create_text_run(doc: &mut XmlDocument, parent: NodeId, text: &str) -> NodeId {
    let run_id = doc.add_child(parent, XmlNodeData::element(W::r()));
    
    let needs_preserve = text.starts_with(' ') || text.ends_with(' ');
    let text_attrs = if needs_preserve {
        vec![XAttribute::new(XName::new("http://www.w3.org/XML/1998/namespace", "space"), "preserve")]
    } else {
        vec![]
    };
    
    let text_element = doc.add_child(run_id, XmlNodeData::element_with_attrs(W::t(), text_attrs));
    doc.add_child(text_element, XmlNodeData::Text(text.to_string()));
    
    run_id
}

pub fn create_paragraph(doc: &mut XmlDocument, parent: NodeId) -> NodeId {
    doc.add_child(parent, XmlNodeData::element(W::p()))
}

pub fn is_revision_element(name: &XName) -> bool {
    name.namespace.as_deref() == Some(W::NS) 
        && matches!(name.local_name.as_str(), "ins" | "del")
}

pub fn is_insertion(name: &XName) -> bool {
    name.namespace.as_deref() == Some(W::NS) && name.local_name == "ins"
}

pub fn is_deletion(name: &XName) -> bool {
    name.namespace.as_deref() == Some(W::NS) && name.local_name == "del"
}

pub fn is_format_change(name: &XName) -> bool {
    name.namespace.as_deref() == Some(W::NS) 
        && matches!(name.local_name.as_str(), "rPrChange" | "pPrChange")
}

static REVISION_ELEMENT_TAGS: &[&str] = &[
    "ins", "del", "rPrChange", "pPrChange", "sectPrChange",
    "tblPrChange", "tblGridChange", "trPrChange", "tcPrChange",
    "cellIns", "cellDel", "cellMerge",
    "customXmlInsRangeStart", "customXmlDelRangeStart",
    "customXmlMoveFromRangeStart", "customXmlMoveToRangeStart",
    "moveFrom", "moveTo", "moveFromRangeStart", "moveToRangeStart",
    "numberingChange",
];

pub fn is_revision_element_tag(local_name: &str) -> bool {
    REVISION_ELEMENT_TAGS.contains(&local_name)
}

pub fn fix_up_revision_ids(doc: &mut XmlDocument, start: NodeId) {
    let mut revision_nodes = Vec::new();
    
    for node_id in doc.descendants(start) {
        if let Some(data) = doc.get(node_id) {
            if let Some(name) = data.name() {
                if name.namespace.as_deref() == Some(W::NS) 
                    && is_revision_element_tag(&name.local_name) 
                {
                    if let Some(attrs) = data.attributes() {
                        if attrs.iter().any(|a| a.name.local_name == "id" && a.name.namespace.as_deref() == Some(W::NS)) {
                            revision_nodes.push(node_id);
                        }
                    }
                }
            }
        }
    }
    
    let mut next_id = 0;
    for node_id in revision_nodes {
        if let Some(data) = doc.get_mut(node_id) {
            if let Some(attrs) = data.attributes_mut() {
                for attr in attrs.iter_mut() {
                    if attr.name.local_name == "id" && attr.name.namespace.as_deref() == Some(W::NS) {
                        attr.value = next_id.to_string();
                        next_id += 1;
                        break;
                    }
                }
            }
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct RevisionCounts {
    pub insertions: usize,
    pub deletions: usize,
    pub format_changes: usize,
}

impl RevisionCounts {
    pub fn total(&self) -> usize {
        self.insertions + self.deletions + self.format_changes
    }
}

/// Count revisions by grouping adjacent revision elements with the same type and attributes
/// This matches the C# GetRevisions algorithm which uses GroupAdjacent
pub fn count_revisions(doc: &XmlDocument, start: NodeId) -> RevisionCounts {
    let mut counts = RevisionCounts::default();
    
    // First pass: collect all paragraphs and their revision children
    // The C# groups by correlation status + revision attributes (author, date)
    // For simplicity, we group adjacent revision elements of the same type with same author/date
    
    count_revisions_in_subtree(doc, start, &mut counts, None);
    
    counts
}

/// Get a grouping key for a revision element (type + author + date, excluding id)
fn get_revision_key(doc: &XmlDocument, node: NodeId) -> Option<String> {
    let data = doc.get(node)?;
    let name = data.name()?;
    
    if name.namespace.as_deref() != Some(W::NS) {
        return None;
    }
    
    let rev_type = match name.local_name.as_str() {
        "ins" => "ins",
        "del" => "del",
        "rPrChange" => "rPrChange",
        "pPrChange" => "pPrChange",
        _ => return None,
    };
    
    // Build key from type + author + date (but NOT id)
    let mut key = rev_type.to_string();
    if let Some(attrs) = data.attributes() {
        for attr in attrs {
            if attr.name.namespace.as_deref() == Some(W::NS) {
                match attr.name.local_name.as_str() {
                    "author" => key.push_str(&format!("|author={}", attr.value)),
                    "date" => key.push_str(&format!("|date={}", attr.value)),
                    _ => {}
                }
            }
        }
    }
    Some(key)
}

/// Recursively count revisions, grouping adjacent siblings with same key
fn count_revisions_in_subtree(
    doc: &XmlDocument,
    node: NodeId,
    counts: &mut RevisionCounts,
    _parent_key: Option<&str>,
) {
    let children: Vec<_> = doc.children(node).collect();
    let mut last_key: Option<String> = None;
    
    for child in children {
        if let Some(key) = get_revision_key(doc, child) {
            // Only count if this is a new revision group (different key from last)
            let is_new_group = last_key.as_ref() != Some(&key);
            
            if is_new_group {
                // Count this as a new revision
                if key.starts_with("ins") {
                    counts.insertions += 1;
                } else if key.starts_with("del") {
                    counts.deletions += 1;
                } else if key.starts_with("rPrChange") || key.starts_with("pPrChange") {
                    counts.format_changes += 1;
                }
            }
            
            last_key = Some(key);
            
            // Don't recurse into revision elements - their content is part of the revision
        } else {
            // Not a revision element - reset the grouping and recurse
            last_key = None;
            count_revisions_in_subtree(doc, child, counts, None);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn revision_settings_default() {
        let settings = RevisionSettings::default();
        assert_eq!(settings.author, "redline-rs");
        assert!(!settings.date_time.is_empty());
    }

    #[test]
    fn revision_id_counter() {
        reset_revision_id_counter(100);
        assert_eq!(get_next_revision_id(), 100);
        assert_eq!(get_next_revision_id(), 101);
        assert_eq!(get_next_revision_id(), 102);
    }

    #[test]
    fn revision_id_starts_at_zero() {
        reset_revision_id_counter(0);
        assert_eq!(get_next_revision_id(), 0);
        assert_eq!(get_next_revision_id(), 1);
        assert_eq!(get_next_revision_id(), 2);
    }

    #[test]
    fn is_revision_element_checks() {
        assert!(is_insertion(&W::ins()));
        assert!(!is_insertion(&W::del()));
        assert!(is_deletion(&W::del()));
        assert!(!is_deletion(&W::ins()));
        assert!(is_format_change(&W::rPrChange()));
        assert!(is_format_change(&W::pPrChange()));
        assert!(!is_format_change(&W::ins()));
    }

    #[test]
    fn create_text_run_works() {
        let mut doc = XmlDocument::new();
        let root = doc.add_root(XmlNodeData::element(W::body()));
        let run = create_text_run(&mut doc, root, "Hello");
        
        assert!(doc.get(run).is_some());
        let run_data = doc.get(run).unwrap();
        assert_eq!(run_data.name().map(|n| &n.local_name), Some(&"r".to_string()));
    }
}
